### ISSUE TITLE
Allow Javarosa forms to be used without an equivalent JSON form

### ISSUE THREAD

**Issue by browndav** [medic/medic-webapp#1091]
Currently, we have two parallel form engines that operate through two different APIs. This is not optimal. We'd like to refactor our forms API to be modular with respect to forms engines. In the ideal outcome, a developer would be able to add support for a third forms engine, by defining methods along the lines of `render(form, data, target)`, `validate_form(form)`, `validate_submission(form, data)`.

The current priority is the Enketo / Javarosa forms system; changes to the legacy JSON-based forms system that aren't absolutely required to avoid regressions could possibly be deferred.

**Comment by browndav** [comment]
Hey @alxndrsn, wondering if you'd be up for taking a look at this one. @mandric put together the current ODK-specific APIs, and can probably help out with questions if we have any.

**Comment by browndav** [comment]
From this morning's call: currently, deploying an ODK/Javarosa form requires that (i) a workalike JSON form be created that exactly matches the structure of the ODK/Javarosa form; and (ii) that the ODK form be uploaded manually using `curl` or similar.

First and foremost, we'd like to break assumption (i) as part of this issue: we should be able to upload an ODK form (through some method, possibly manual) and use it to view/submit data _without_ a workalike JSON form being present. This may break code that assumes form definitions are available in the design document, such as the `/add` update handler. If this happens, we should move it to `medic-api`.

Once this is working, we can file a separate issue to clean up the forms API, and/or refactor the legacy JSON form handling code to fit the new API. This change will also open the door to accepting XML-formatted data from Medic Collect and/or Enketo (currently, Medic Collect sends a serialized SMS inside of an HTTP payload).

**Comment by browndav** [comment]
Update from this morning: this issue is intended to clear the way for in-app creation/modification of reports using Enketo. We need to be able to store reports ("records") as XML for Enketo to understand them, which in turn means that programmatic accesses to these reports (e.g. examining the patient_id field) can no longer directly access properties, and need to be refactored to use a form-type-aware accessor. Think `form(definition, report).get_field('patient_id')`, but with different implementations for different types of form definitions.

**Comment by alxndrsn** [comment]
Form submission and display in the `Records` tab is working on https://github.com/alxndrsn/medic-webapp/tree/1091-rewrite-forms-storage.  This is branched from https://github.com/alxndrsn/medic-webapp/tree/1086_embed-enketo-core-with-browserify, and is not a huge change.

I will look further into:
* editing forms with enketo
* the `Analytics` tab, and how it's relevant to this work

For now though, any feedback on approach so far would be appreciated, although this may be blocked by https://github.com/medic/medic-webapp/issues/1086.
